File: data/data_manager.py
import os
import json

class DataManager:
    def __init__(self, main):
        from data.objects import character as ch, enemy as en, bullet

        self.main = main
        self.data_schema: dict = {
            'class': ch.CharacterClass.from_json,
            'enemy': en.Enemy.from_json,
            'bullet': bullet.Bullet.from_json,
        }
        self.data: dict = {key: {} for key in self.data_schema.keys()}

    def get_character_class(self, id: str):
        return self.data['class'].get(id, None)

    def get_enemy(self, id: str):
        return self.data['enemy'].get(id, None)

    def get_bullet(self, id: str):
        return self.data['bullet'].get(id, None)

    def load(self):        
        base = 'resources/data/'

        if not os.path.exists(base):
            print(f"Directory '{base}' does not exist.")
            return

        for folder, parser in self.data_schema.items():
            folder_path = os.path.join(base, folder)

            if not os.path.exists(folder_path):
                print(f"Folder '{folder_path}' does not exist, skipping.")
                continue

            for filename in os.listdir(folder_path):
                if filename.endswith(".json"):
                    file_path = os.path.join(folder_path, filename)
                    try:
                        with open(file_path, "r", encoding="utf-8") as file:
                            obj = parser(json.load(file))
                            self.data[folder][obj.id] = obj
                    except json.JSONDecodeError as e:
                        print(f"Error parsing {filename} in {folder}: {e}")

File: server/src/main/java/dev/creoii/chaos/DataManager.java
package dev.creoii.chaos;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.JsonReader;
import dev.creoii.chaos.data.JsonParsing;
import dev.creoii.chaos.entity.BulletEntityType;
import dev.creoii.chaos.entity.EnemyEntityType;
import dev.creoii.chaos.entity.LootDropEntityType;
import dev.creoii.chaos.entity.character.CharacterClass;
import dev.creoii.chaos.item.ServerItem;
import dev.creoii.chaos.util.Identifiable;
import dev.creoii.chaos.util.Parser;

import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.Map;

public class DataManager {
    private final Map<String, Parser> schema;
    private final Map<String, Map<String, Identifiable>> data;

    public DataManager() {
        schema = new HashMap<>();
        schema.put("class", fileHandle -> JsonParsing.parseCharacterClass(fileHandle.nameWithoutExtension(), new JsonReader().parse(fileHandle)));
        schema.put("item", fileHandle -> JsonParsing.parseItem(fileHandle.nameWithoutExtension(), new JsonReader().parse(fileHandle)));
        schema.put("enemy", fileHandle -> JsonParsing.parseEnemyEntityType(fileHandle.nameWithoutExtension(), new JsonReader().parse(fileHandle)));
        schema.put("bullet", fileHandle -> JsonParsing.parseBulletEntityType(fileHandle.nameWithoutExtension(), new JsonReader().parse(fileHandle)));
        schema.put("loot_drop", fileHandle -> JsonParsing.parseLootDropEntityType(fileHandle.nameWithoutExtension(), new JsonReader().parse(fileHandle)));

        data = new HashMap<>();
        for (String key : schema.keySet()) {
            data.put(key, new HashMap<>());
        }
    }

    @Nullable
    public CharacterClass getCharacterClass(String id) {
        return (CharacterClass) data.get("class").getOrDefault(id, null);
    }

    @Nullable
    public EnemyEntityType getEnemy(String id) {
        return (EnemyEntityType) data.get("enemy").getOrDefault(id, null);
    }

    @Nullable
    public BulletEntityType getBullet(String id) {
        return (BulletEntityType) data.get("bullet").getOrDefault(id, null);
    }

    @Nullable
    public LootDropEntityType getLootDrop(String id) {
        return (LootDropEntityType) data.get("loot_drop").getOrDefault(id, null);
    }

    @Nullable
    public ServerItem getItem(String id) {
        return (ServerItem) data.get("item").getOrDefault(id, null);
    }

    public void load() {
        FileHandle baseDir = Gdx.files.internal("data");

        if (!baseDir.exists()) {
            Gdx.app.log(DataManager.class.getSimpleName(), "Directory 'data/' does not exist.");
            return;
        }

        for (Map.Entry<String, Parser> entry : schema.entrySet()) {
            String folder = entry.getKey();
            Parser parser = entry.getValue();

            FileHandle folderHandle = baseDir.child(folder);
            if (!folderHandle.exists()) {
                Gdx.app.log(DataManager.class.getSimpleName(), "Folder '" + folderHandle.path() + "' does not exist, skipping.");
                continue;
            }

            for (FileHandle file : folderHandle.list("json")) {
                try {
                    Identifiable obj = parser.parse(file);
                    data.get(folder).put(obj.id(), obj);
                } catch (Exception e) {
                    Gdx.app.error(DataManager.class.getSimpleName(), "Error parsing " + file.name() + " in '/" + folder + "': " + e);
                }
            }
        }
    }
}
